### PR TITLE
[script][status-monitor] Refactor: bug fixes, SOLID decomposition, SQLite storage

### PR DESCRIPTION
**Pull request by MahtraDR**
## Summary

Refactors status-monitor.lic from a monolithic `GameFilter` class into focused, testable modules and fixes several bugs in the existing implementation.

### Bug fixes
- **Gdk::RBGA typo** -- crashed health bar rendering when HP dropped below 75%
- **Kernel#open -> File.open** -- closed a command injection vector in Marshal file loading
- **File handle leaks** -- Marshal.load calls now properly close file handles
- **similarity_scrub mutation** -- no longer mutates its argument via gsub!
- **@room_players unbounded growth** -- resets on room change instead of accumulating forever
- **@filter_strings busy-wait** -- retries with backoff instead of spinning on nil
- **Double Slack notification** -- deduplicates alert delivery
- **Duplicate @non_useful_tags** -- removed repeated `death`, `percWindow` entries

### Structural changes
- Decomposes `GameFilter` into focused classes: `MessageStore`, `MessageFilter`, `SpamDetector`, `AlertHandler`, `CommandDetector`, `Monitor`
- Replaces Marshal `.dat` persistence with SQLite (WAL mode) for the seen-messages corpus
- Auto-migrates existing `.dat` files on first run, then renames to `.dat.migrated`
- Switches Slack delivery from the shared `register_slackbot`/`send_slackbot_message` globals to a dedicated, lazily-constructed `Lich::DragonRealms::SlackBot` instance (delivery delegates reconnection to `direct_message`)

### New script
- Adds `status-monitor-import.lic` for bulk-importing game session logs (`.log` and `.log.gz`) into the SQLite corpus. Resumable, tracks processed files.

### Review follow-ups
- **Slack alert delivery** -- `AlertHandler#send_slack` no longer gates on `initialized?`; that gate permanently suppressed delivery whenever the first connection attempt failed (e.g. lnet not up at login), because `direct_message`'s own reconnect path was never reached
- **Lazy Slack init** -- the `SlackBot` is constructed on the first alert instead of at script start, so the status window is no longer blocked behind lnet/users.list startup
- **Multi-instance log import** -- `find_log_files` globs every `*-<Character>` log directory (DR/DRT/DRX/GSIV) instead of hardcoding the `DR-` prefix
- **`--limit` semantics** -- the limit now caps not-yet-imported files, so a resumed limited run makes forward progress instead of re-selecting already-imported files
- **`MessageFilter#clean`** -- no longer mutates its argument (tag stripping returns a new String; safe against frozen input)
- **Command auto-execution** -- destructive/session-ending verbs (QUIT, SELL, DROP, DUMP, EMPTY, WITHDRAW, RELEASE, GET, PUT, KILL) are denylisted from `CommandDetector`
- **Docs/rubocop** -- documented `last_player_line`, named the recent-cache TTL constant, and kept the adversarial spec inputs ASCII-only so the repo cops pass
- Test coverage added for each of the above (169 examples, rubocop clean)

## Test plan
- [ ] Run `;status-monitor` on a character with an existing `.dat` file -- verify auto-migration to SQLite and `.dat.migrated` rename
- [ ] Run `;status-monitor` on a fresh character with no prior data -- verify clean SQLite database creation
- [ ] Verify health bar colors update correctly at HP thresholds (especially < 75%)
- [ ] Confirm Slack notifications fire once per alert, not duplicated
- [ ] Confirm a Slack alert is delivered even when the character logs in before lnet is connected
- [ ] Run `;status-monitor-import` to populate corpus from logs, then re-run to verify resumability
- [ ] Verify `--limit=N` imports N not-yet-imported files on a resumed run
- [ ] Verify a character with logs on multiple instances (e.g. DR- and DRT-) imports all of them
- [ ] Verify `--reset` flag on import clears tracking and re-processes all files

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added per-character log ingestion (plain and gzip) into a SQLite-backed message corpus, including deduping, `--reset`, and progress/completion reporting.
  * Reworked the status monitoring pipeline to use an SQLite `MessageStore`, improved message cleaning/similarity normalization, spam alerting, and embedded command detection/execution.
* **Bug Fixes**
  * Fixed health bar color parsing.
* **Tests**
  * Added extensive RSpec coverage for log import behavior, cleaning/scrubbing rules, database operations, command detection, spam detection, and alert handling.

### CODE REVIEW

File: status-monitor-import.lic [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  XML_TAG_PATTERN = /<[^>]+>/
+  REGET_MARKER = /<!-- Above contents from reget/
+  CURRENCY_SCRUB = [/[0-9]+/, /(kronars|lirums|dokoras)/i].freeze
+
+  # Lines matching these patterns are not useful for the seen corpus.
+  SKIP_PATTERNS = [
+    /^$/,
+    /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}/,  # bare timestamp lines (session headers)
+    /^>/, # upstream commands
+  ].freeze
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor-import.lic:43]: Fixed in 5f9312bd. Removed the timestamp regex from `SKIP_PATTERNS` entirely rather than end-anchoring. The `TIMESTAMP_PATTERN` sub + empty check handles bare timestamp headers naturally, avoiding a fragile regex that must account for varying timezone formats (`NZDT`, `EST`, `+13:00`).

Verified against real log files using custom `%F %T %Z` format. Covered by spec: `does not skip timestamped content lines (regression test)`.

File: status-monitor-import.lic [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  # @param db [SQLite3::Database] database connection
+  # @param lines [Array<String>] lines to insert
+  # @return [Integer] number of lines inserted (approximate)
+  def self.flush_batch(db, lines)
+    db.transaction do
+      lines.each do |line|
+        db.execute('INSERT OR IGNORE INTO seen_messages (line_hash, source) VALUES (?, ?)', line, 'import')
+      end
+    end
+    lines.size
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor-import.lic:171]: Addressed in 030293ab. Changed progress labels from "lines" to "lines scanned" since `INSERT OR IGNORE` silently skips duplicates. The final summary already correctly reports `count_after - count_before` for actual new unique entries. Did not switch to `db.changes` per-row counting -- the overhead of summing per-INSERT in a 10k batch is not worth the marginal accuracy improvement on an intermediate progress line.

File: status-monitor-import.lic [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  # @param character_name [String] character to import for
+  # @param limit [Integer, nil] max files to process
+  # @param reset [Boolean] if true, re-process all files
+  # @return [void]
+  def self.run(character_name, limit: nil, reset: false)
+    db = open_database(character_name)
+
+    if reset
+      db.execute('DELETE FROM import_log')
+      echo 'Import log cleared -- will re-process all files.'
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor-import.lic]: Fixed in 5f9312bd. Added `DELETE FROM seen_messages WHERE source = 'import'` before clearing `import_log`. This preserves `live` and `migration` sourced rows while allowing a true re-derivation of the import corpus. Covered by spec: `deletes import-sourced rows but preserves live rows` and `preserves migration-sourced rows`.

File: status-monitor-import.lic [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    total_lines = 0
+    skipped = 0
+    processed = 0
+    start_time = Time.now
+
+    count_before = db.get_first_value('SELECT COUNT(*) FROM seen_messages').to_i
+
+    echo "Found #{total_files} log files for #{character_name}."
+
+    files.each_with_index do |file, idx|
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor-import.lic]: Fixed by `rubocop -A` in the initial cleanup pass. Changed to `_idx` prefix.

File: status-monitor.lic [hunk 5]
```diff
@@ -17,294 +17,554 @@ args = parse_args(arg_definitions)
...
+            fput(match.downcase)
+            pause 5
+          end
+
+      line.split(' ').each do |token|
+        next unless token =~ /[_~=-]/
+
+        cleaned = token.gsub(/_|~|-|=/, '')
+        fput(cleaned.downcase) if VALID_COMMANDS.include?(cleaned.upcase)
+      end
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor.lic]: Fixed in 5f9312bd. Both scanners now collect into a `Set`, then a single loop executes each unique command once. Verified with `J_U_M_P` input -- confirmed it matched both scanners but executes only once after dedup. Covered by spec: `deduplicates commands that match both scanners`.

File: status-monitor.lic [hunk 6]
```diff
@@ -17,294 +17,554 @@ args = parse_args(arg_definitions)
...
+    def process(line)
+      @store.migrate_recent
+      cleaned = @filter.clean(line)
+      return false if cleaned.nil?
+      return false if @filter.filtered?(cleaned)
+
+      scrubbed = @filter.similarity_scrub(cleaned)
+      return false unless @store.unseen?(scrubbed)
+
+      alert = @detector.check(scrubbed)
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor.lic]: Fixed in 5f9312bd. Moved `@detector.check(scrubbed)` before `@store.unseen?(scrubbed)` in `Monitor#process`, restoring the upstream ordering where `spam_check` runs before the recent-cache guard. The detector now sees all cleaned+scrubbed lines including repeats, so `counts.values.max` can exceed the threshold. Covered by spec: `detector runs before unseen? gate (spam detection regression test)`.

File: status-monitor.lic [hunk 7]
```diff
@@ -17,294 +17,554 @@ args = parse_args(arg_definitions)
...
+
+    private
+
+    # @return [Array<Regexp>] compiled filter patterns from data file
+    def load_filter_strings
+      retries = 0
+      filters = nil
+      while filters.nil?
+        data = get_data('filters')
+        filters = data['filter_strings']&.map { |string| /#{string}/ }
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor.lic:619]: No fix needed. `get_data` returns an `OpenStruct` (even in its rescue path), and `OpenStruct#[]` returns `nil` for missing keys without raising. The safe navigation `data['filter_strings']&.map` on line 548 already handles the nil case, and the retry loop with `pause 1` + max 10 attempts handles transient load failures. Verified by tracing through the `SetupFiles#get_data` -> `transform_data` -> `OpenStruct.new` chain in dependency.lic.

File: status-monitor.lic [hunk 8]
```diff
@@ -17,294 +17,554 @@ args = parse_args(arg_definitions)
...
-    end
-    if filter.spam_line
-      filter.spam_line.dup
-      filter.spam_line = nil
+    if monitor.process(line)
+      write_debug_log(back, line, monitor.last_player_line) if args.debug
     end
+    monitor.consume_spam_line
   end
 end
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor.lic:671]: Fixed in 030293ab. Registered `before_dying { monitor.shutdown }` immediately after `Monitor.new`, before the nowindow branch. Both headless and GTK modes now get clean DB flush + close on exit. The GTK-specific `before_dying` block was kept separately for window state persistence only, with `monitor.save` removed (handled by `shutdown`).

**Comment by MahtraDR** [comment]
Re: out-of-diff comment about `script.gets?` returning nil and `line.dup` crashing in GTK debug mode (line 666):

No fix needed. `nil.dup` returns `nil` without raising in Ruby 2.4+ (lich requires Ruby 3.x). The nil then propagates into `monitor.process(nil)` where `MessageFilter#clean` returns nil early on the `line.nil?` check (line 199). This is the same behavior as upstream (line 380). Verified with `ruby -e "puts nil.dup.inspect"` -> `nil`.

**Comment by MahtraDR** [comment]
@coderabbitai - update review on this pr to include all commits.

File: status-monitor-import.lic [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+               File.open(path, 'r')
+             end
+
+    reader.each_line do |raw_line|
+      cleaned = clean_line(raw_line)
+      next unless cleaned
+
+      scrubbed = similarity_scrub(cleaned)
+      next if scrubbed.empty?
+
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor-import.lic]: Fixed in 00b2f498 -- added `.strip` before `.empty?` check so whitespace-only strings left by currency/digit removal are skipped.

**Comment by MahtraDR** [comment]
Addressing @CodeRabbit review (4307953260):

**status-monitor-import.lic:138-140 (whitespace after similarity_scrub) -- Fixed.**
Added `.strip` before `.empty?` so strings like `"  "` left after currency/digit removal are properly skipped. Replied inline.

**spec/status_monitor_spec.rb:619-623 (unused `stub_get_data`) -- Fixed.**
Removed the dead helper. It was never called (all tests use `Object.send(:define_method, :get_data)` inline), and it was also broken -- bare `define_method` would define on the example group, not on `Object` where `Monitor` needs it.

Both fixes in 00b2f498.

**Comment by MahtraDR** [comment]
@coderabbitai review this pr again taking all commits into account together.

File: status-monitor-import.lic [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  ]
+]
+
+args = parse_args(arg_definitions)
+
+module StatusMonitorImport
+  TIMESTAMP_PATTERN = /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}[^:]*: /
+  XML_TAG_PATTERN = /<[^>]+>/
+  REGET_MARKER = /<!-- Above contents from reget/
+  CURRENCY_SCRUB = [/[0-9]+/, /(kronars|lirums|dokoras)/i].freeze
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor-import.lic:37]: Good catch. Added `BARE_TIMESTAMP_PATTERN` to catch timestamp-only session headers (both `NZDT`-style and `+13:00`-style) and filter them in `clean_line`. Updated specs to verify they're now rejected.

File: status-monitor-import.lic [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    echo "Found #{total_files} log files for #{character_name}."
+
+    files.each_with_index do |file, _idx|
+      if already_imported?(db, file)
+        skipped += 1
+        next
+      end
+
+      lines = import_file(db, file)
+      record_import(db, file, lines)
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor-import.lic]: Good catch. Wrapped `import_file` in a begin/rescue for `Zlib::GzipFile::Error`, `EOFError`, and `SystemCallError`. Corrupt or truncated files are now logged and skipped, with an error count in the final summary.

File: status-monitor.lic [hunk 12]
```diff
@@ -17,294 +17,595 @@ args = parse_args(arg_definitions)
...
-    @responses = ["'Hmmm?", "'Yes", "'Ok?"].shuffle
-    @settings = get_settings
-    @recent_seen = []
-    @frequency_buffer = []
-    @quit_on_flag = @settings.quit_on_status_warning
-    @unique_line_threshold = @settings.unique_line_threshold
-    @line_frequency_threshold = @settings.line_frequency_threshold
-    @line_similarity_percentage = @settings.line_similarity_percentage
-  end
+require 'sqlite3'
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor.lic:20]: Not applicable -- `Set` has been a core class (no require needed) since Ruby 3.2, and both dr-scripts and lich-5 target Ruby 4.0+. The `require 'set'` was intentionally removed in d0486747 because Rubocop flagged it as redundant.

File: status-monitor.lic [hunk 13]
```diff
@@ -17,294 +17,595 @@ args = parse_args(arg_definitions)
...
+      count = 0
+      @db.transaction do
+        data.each_key do |line|
+          insert_line(line.to_s, 'migration')
+          count += 1
+        end
+      end
+      File.rename(@legacy_dat, "#{@legacy_dat}.migrated")
+      backup = "backup/seen_messages_#{File.basename(@legacy_dat, '.dat')}.bak"
+      File.rename(backup, "#{backup}.migrated") if File.exist?(backup)
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor.lic]: Good catch. `@legacy_dat` is `"seen_messages_Mahtra.dat"`, so `File.basename(@legacy_dat, '.dat')` returns `"seen_messages_Mahtra"` -- prepending another `"seen_messages_"` doubles the prefix. Fixed to `"backup/#{File.basename(@legacy_dat, '.dat')}.bak"`. The spec had the same bug (test was green because it created the file at the same wrong path); fixed there too.

File: status-monitor.lic [hunk 14]
```diff
@@ -17,294 +17,595 @@ args = parse_args(arg_definitions)
...
+
+      scrubbed = @filter.similarity_scrub(cleaned)
+
+      alert = @detector.check(scrubbed)
+      if alert
+        @alerter.fire(alert[:line], alert[:counts])
+        @spam_line = alert[:line]
+      end
+
+      @store.unseen?(scrubbed)
```
[the remark this answers is not in the document]
Comment (MahtraDR) [reply status-monitor.lic]: Good catch. Added `.strip` and empty guard after `similarity_scrub` in `Monitor#process` (matching the pattern already used in `status-monitor-import.lic`). Also tightened `MessageStore#unseen?` to reject whitespace-only lines with `line.strip.empty?` for defense-in-depth. Updated specs.

**Comment by MahtraDR** [comment]
@coderabbitai review commit 1553e94b and commit b2fa2d6a

These two commits address all findings from your previous review:

**1553e94b** - Source fixes:
- Bare timestamp headers now filtered via `BARE_TIMESTAMP_PATTERN`
- Corrupt `.gz` files rescued and skipped in import loop
- Backup migration path fixed (was doubling `seen_messages_` prefix)
- Whitespace-only scrubbed lines now guarded in both `Monitor#process` and `MessageStore#unseen?`
- CommandDetector strips non-ASCII before scanning (closes zero-width char evasion)
- SpamDetector uses time-windowed counts instead of flushable 20-entry buffer
- MessageFilter uses proper stream state tracking (like `xmlparser.rb`) for all filtered streams

**b2fa2d6a** - Comprehensive edge case, boundary, and adversarial specs (36 new tests, 152 total)

**Comment by MahtraDR** [comment]
@coderabbitai Both nitpicks from your review are fixed in 509d1744:

1. **Unused `let(:original_dir)`** -- removed from `status_monitor_spec.rb`
2. **Orphaned LICH_DIR temp directory** -- added `after(:suite)` cleanup hook in `status_monitor_import_spec.rb`

**Comment by MahtraDR** [comment]
@coderabbitai review this PR in full

**Comment by MahtraDR** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>
